### PR TITLE
Simplifying mxnet.gluon.block APIs

### PR DESCRIPTION
**Pull request by acphile**
## Description ##
A draft PR for issue https://github.com/apache/incubator-mxnet/issues/18412

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by acphile** [comment]
@mxnet-bot run ci [centos-gpu]

File: tests/python/unittest/test_gluon_estimator.py [hunk 1]
```diff
@@ -394,24 +388,6 @@ def test_val_net():
                     val_loss=val_loss,
                     val_net=val_net)
 
-    with pytest.raises(RuntimeError):
-        est.fit(train_data=dataloader,
-                val_data=dataloader,
-                epochs=num_epochs)
-
-    ''' test weight sharing of sequential networks with namescope '''
```
[the remark this answers is not in the document]
Comment (acphile) [reply tests/python/unittest/test_gluon_estimator.py:402]: Actually test for new method of weight sharing exists. Originally, there are two tests: "test weight sharing of sequential networks with namescope" and "test weight sharing of sequential networks without namescope". Since in new API, we don't have to define children block inside "with net.name_scope()", I remove one of the tests. You can see in "def _get_test_network(params=None)", I maintain the original test for weight sharing.

**Comment by leezu** [comment]
@mxnet-bot run ci [website, unix-cpu]

File: python/mxnet/gluon/block.py [hunk 2]
```diff
@@ -542,61 +460,60 @@ def load_parameters(self, filename, ctx=None, allow_missing=False,
...
+
+    def load_dict(self, param_dict, filename=None, ctx=None, allow_missing=False,
+                  ignore_extra=False, cast_dtype=False, dtype_source="current"):
+        """Load parameters from dict
+
+        Parameters
+        ----------
+        param_dict : dict
+            Dictionary containing model parameters
+        filename : str, default None
```
Comment (leezu) [inline python/mxnet/gluon/block.py:476]: When making `load_dict` a public function (instead of internal), it shouldn't take a `filename` argument. The use of filename here is just due to how you implement `load_parameters`.

File: python/mxnet/gluon/parameter.py [hunk 3]
```diff
@@ -431,7 +437,10 @@ def initialize(self, init=None, ctx=None, default_init=initializer.Uniform(),
             and :py:meth:`Parameter.init` are ``None``.
         force_reinit : bool, default False
             Whether to force re-initialization if parameter is already initialized.
-
+        structural_name : str, default ""
+            The structural name for the parameter in the block.
+            The value would be accessed in InitDesc.attrs['structure'] by self-defined initializers.
+            Users may want to initialize parameters based on the block's structure
```
Comment (leezu) [inline python/mxnet/gluon/parameter.py:443]: Unused?

**Comment by sxjscience** [comment]
@leezu This is a major API upgrade and we will need to rewrite + test the majority of gluonnlp/numpy: https://github.com/dmlc/gluon-nlp/tree/numpy

**Comment by leezu** [comment]
@sxjscience Yes.. Though I believe this is unavoidable. Not even the D2L book manages to introduce the existing API correctly.. For example, https://d2l.ai/chapter_deep-learning-computation/model-construction.html is plain wrong in MXNet 1..

**Comment by RuRo** [comment]
**(Disclamer)** I didn't look too deeply into the reasons for these changes, so I may be wrong, but... **(Disclamer)**

In my opinion, the distinction between the "unique/unreadable" parameter names and the "human-readable" parameter names is quite confusing. It's not obvious, where should you expect which kind of parameter name.

For example, currently `.collect_params()` seems to return human-readable names as keys, but the parameter values themselves have "unique" `.name`s.

Exceptions, errors and warnings (for example the "Parameter has not been initialized." error) currently seem to print the "unique" parameter name, which is really unhelpful for debugging purposes, since it doesn't tell you, where is the uninitialized parameter located.

I am not very familiar with the inner workings of Gluon Parameters, but it really feels like there has to be a better way.

**Comment by leezu** [comment]
@RuRo Thanks for pointing out the problem. Generally the structural name should be the "public" or user-visible name. The UUID is mainly to workaround some buggy MXNet operators that have global state based on symbol names. So overall the idea is to get rid of any global topological information `parameter.name`. Let's open an issue to track the problems where the internal name is exposed and improve the behaviour.

**Comment by acphile** [comment]
For warnings in method `intialize()`, I pass parameter `structural_name` which is a readable name based on model structure, and we can replace `self.name` with `self._structural_name`. But I am not sure about other methods of `Parameter`: if those method are executed after `initialize()`, then we can do similar replacements accordingly.
